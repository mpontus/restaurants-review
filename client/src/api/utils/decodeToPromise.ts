import * as t from "io-ts";
import { failure } from "io-ts/lib/PathReporter";

/**
 * Validate input against io-ts schema and return result as a promise
 */
export const decodeToPromise = <A, O, I>(
  validator: t.Type<A, O, I>,
  input: I
): Promise<A> =>
  validator
    .decode(input)
    .fold(
      errors => Promise.reject(new Error(failure(errors).join("\n"))),
      value => Promise.resolve(value)
    );
