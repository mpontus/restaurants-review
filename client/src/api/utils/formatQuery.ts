import qs from "querystring";

/**
 * Filter undefined values from object
 */
const filterObject = (obj: Record<string, any>): object =>
  Object.keys(obj).reduce(
    (acc, key) => (obj[key] === undefined ? acc : { ...acc, [key]: obj[key] }),
    {}
  );

/**
 * Format query string from object after filtering undefined values
 */
export const formatQuery = (query: object) => qs.stringify(filterObject(query));
