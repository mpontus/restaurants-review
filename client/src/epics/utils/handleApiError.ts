import { of, OperatorFunction } from "rxjs";
import { catchError } from "rxjs/operators";
import { RequestError } from "../../models/RequestError";

/**
 * Helper function to handle API error
 *
 * Wraps error caught from source observable into RequestError and
 * passes it onto specified action creator before merging it into the
 * result stream.
 */
export const handleApiError = <T, R>(
  failureActionCreator: (e: any) => R
): OperatorFunction<T, T | R> =>
  catchError(error =>
    of(failureActionCreator(RequestError.fromApiError(error)))
  );
