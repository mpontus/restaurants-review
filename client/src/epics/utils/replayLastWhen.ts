import { merge, Observable, OperatorFunction } from "rxjs";
import { withLatestFrom } from "rxjs/operators";

/**
 * Replays the last item emitted by the source observable, whenever
 * selector emits.
 */
export const replayLastWhen = <T>(
  selector$: Observable<any>
): OperatorFunction<T, T> => source$ =>
  merge(source$, selector$.pipe(withLatestFrom(source$, (_, val) => val)));
