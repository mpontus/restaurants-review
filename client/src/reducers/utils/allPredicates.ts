/**
 * Helper type
 */
type Predicate<T, T1 extends T> = (input: T) => input is T1;

/**
 * All Predicates
 *
 * Creates a predicate which matches all of the passed predicates.
 */
export function allPredicates<T, T1 extends T>(
  predicate1: Predicate<T, T1>,
  ...rest: Array<(value: T1) => boolean>
): Predicate<T, T1>;

export function allPredicates<T, T1 extends T, T2 extends T1>(
  predicate1: Predicate<T, T1>,
  predicate2: Predicate<T1, T2>,
  ...rest: Array<(value: T2) => boolean>
): Predicate<T, T2>;

export function allPredicates<T, T1 extends T, T2 extends T1, T3 extends T2>(
  predicate1: Predicate<T, T1>,
  predicate2: Predicate<T1, T2>,
  predicate3: Predicate<T2, T3>,
  ...rest: Array<(value: T3) => boolean>
): Predicate<T, T3>;

export function allPredicates(...predicates: Array<(value: any) => boolean>) {
  return (input: any) => predicates.every(predicate => predicate(input));
}
