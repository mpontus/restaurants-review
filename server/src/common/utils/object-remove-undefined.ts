/**
 * Remove undefined values from an object
 */
export const objectRemoveUndefined = <T>(
  // tslint:disable-next-line:no-any
  obj: Partial<T> & { [K in string]?: any },
): Partial<T> =>
  Object.keys(obj).reduce(
    (acc, key) =>
      obj[key] === undefined ? acc : Object.assign(acc, { [key]: obj[key] }),
    {},
  );
