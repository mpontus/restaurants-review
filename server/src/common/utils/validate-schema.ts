// tslint:disable:promise-function-async
import * as t from 'io-ts';
import { failure } from 'io-ts/lib/PathReporter';

/**
 * Validate Schema Error
 *
 * Error dispatched when schema validation fails.
 */
export class ValidateSchemaError extends Error {}

/**
 * Validate object against the given schema
 *
 * Used to achieve type-safety on external objects, which shape can
 * not be trusted. Dispatches {@link ValidateSchemaError} when schema
 * validation fails.
 */
export const validateSchema = <A, O, I>(
  validator: t.Type<A, O, I>,
  input: I,
): Promise<A> =>
  validator
    .decode(input)
    .fold(
      errors =>
        Promise.reject(new ValidateSchemaError(failure(errors).join('\n'))),
      value => Promise.resolve(value),
    );
