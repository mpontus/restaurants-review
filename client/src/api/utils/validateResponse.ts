import { AxiosResponse } from "axios";
import * as t from "io-ts";
import { decodeToPromise } from "./decodeToPromise";

/**
 * Validate axios response against io-ts schema and cast it to
 * corresponding type.
 */
export const validateResponse = <A, O, I>(schema: t.Type<A, O, I>) => (
  response: AxiosResponse<I>
): Promise<A> => decodeToPromise(schema, response.data);
