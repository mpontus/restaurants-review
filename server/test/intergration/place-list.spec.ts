import supertest from 'supertest';
import { initApp } from '../utils/init-app';
import { resetDb } from '../utils/reset-db';

let nestApp: any;

beforeAll(async () => {
  nestApp = await initApp();
});

afterAll(() => nestApp.close());

beforeEach(() => resetDb(nestApp));

describe('List Places', () => {
  const seed = require('../seed/multiple-places');

  beforeEach(() => seed.run(nestApp));

  describe('without any criteria', () => {
    it('should succeed', async () => {
      const response = await supertest(nestApp.getHttpServer())
        .get(`/places`)
        .expect(200);

      expect(response.body).toMatchSnapshot();
    });
  });

  describe('when limit is provided', () => {
    it('should succeed', async () => {
      const response = await supertest(nestApp.getHttpServer())
        .get(`/places?take=2`)
        .expect(200);

      expect(response.body).toMatchSnapshot();
    });
  });

  describe('when offset is provided', () => {
    it('should succeed', async () => {
      const response = await supertest(nestApp.getHttpServer())
        .get(`/places?skip=3`)
        .expect(200);

      expect(response.body).toMatchSnapshot();
    });
  });

  describe('when limit is not a number', () => {
    it('should fail', async () => {
      const response = await supertest(nestApp.getHttpServer())
        .get(`/places?take=foo`)
        .expect(400);

      expect(response.body).toMatchSnapshot();
    });
  });

  describe('when limit is negative', () => {
    it('should fail', async () => {
      const response = await supertest(nestApp.getHttpServer())
        .get(`/places?take=-1`)
        .expect(400);

      expect(response.body).toMatchSnapshot();
    });
  });

  describe('when limit is too large', () => {
    it('should fail', async () => {
      const response = await supertest(nestApp.getHttpServer())
        .get(`/places?take=9000`)
        .expect(400);

      expect(response.body).toMatchSnapshot();
    });
  });

  describe('when offset is not a number', () => {
    it('should fail', async () => {
      const response = await supertest(nestApp.getHttpServer())
        .get(`/places?skip=foo`)
        .expect(400);

      expect(response.body).toMatchSnapshot();
    });
  });

  describe('when offset is negative', () => {
    it('should fail', async () => {
      const response = await supertest(nestApp.getHttpServer())
        .get(`/places?skip=-1`)
        .expect(400);

      expect(response.body).toMatchSnapshot();
    });
  });
});
