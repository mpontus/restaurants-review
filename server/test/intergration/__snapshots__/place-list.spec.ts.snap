// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List Places when limit is negative should fail 1`] = `
Object {
  "error": "Bad Request",
  "message": Array [
    Object {
      "children": Array [],
      "constraints": Object {
        "min": "take must not be less than 0",
      },
      "property": "take",
      "target": Object {
        "skip": 0,
        "take": -1,
      },
      "value": -1,
    },
  ],
  "statusCode": 400,
}
`;

exports[`List Places when limit is not a number should fail 1`] = `
Object {
  "error": "Bad Request",
  "message": Array [
    Object {
      "children": Array [],
      "constraints": Object {
        "max": "take must not be greater than 50",
        "min": "take must not be less than 0",
      },
      "property": "take",
      "target": Object {
        "skip": 0,
        "take": null,
      },
      "value": null,
    },
  ],
  "statusCode": 400,
}
`;

exports[`List Places when limit is provided should succeed 1`] = `
Object {
  "items": Array [
    Object {
      "address": "1187 Weskut Ridge",
      "id": "d3973942-133b-55e4-920e-7b0033ab9d38",
      "rating": 0,
      "title": "The Noodle Farmer",
    },
    Object {
      "address": "375 Uciwa Loop",
      "id": "55b62c9d-d3f6-5973-80cc-ca2744b0c1a6",
      "rating": 0,
      "title": "The Honor Goat",
    },
  ],
  "total": 20,
}
`;

exports[`List Places when limit is too large should fail 1`] = `
Object {
  "error": "Bad Request",
  "message": Array [
    Object {
      "children": Array [],
      "constraints": Object {
        "max": "take must not be greater than 50",
      },
      "property": "take",
      "target": Object {
        "skip": 0,
        "take": 9000,
      },
      "value": 9000,
    },
  ],
  "statusCode": 400,
}
`;

exports[`List Places when offset is negative should fail 1`] = `
Object {
  "error": "Bad Request",
  "message": Array [
    Object {
      "children": Array [],
      "constraints": Object {
        "min": "skip must not be less than 0",
      },
      "property": "skip",
      "target": Object {
        "skip": -1,
        "take": 10,
      },
      "value": -1,
    },
  ],
  "statusCode": 400,
}
`;

exports[`List Places when offset is not a number should fail 1`] = `
Object {
  "error": "Bad Request",
  "message": Array [
    Object {
      "children": Array [],
      "constraints": Object {
        "min": "skip must not be less than 0",
      },
      "property": "skip",
      "target": Object {
        "skip": null,
        "take": 10,
      },
      "value": null,
    },
  ],
  "statusCode": 400,
}
`;

exports[`List Places when offset is provided should succeed 1`] = `
Object {
  "items": Array [
    Object {
      "address": "1253 Pusi Road",
      "id": "381424b3-ba66-5444-8d7a-a452071dc3b3",
      "rating": 0,
      "title": "The Bitter Window",
    },
    Object {
      "address": "1528 Aceer Way",
      "id": "36241277-6054-5955-a823-6a0706710235",
      "rating": 0,
      "title": "The Fire Grove",
    },
    Object {
      "address": "826 Cuzmav Ridge",
      "id": "f2ff6a02-8f86-51f7-bd0e-50099e450c27",
      "rating": 0,
      "title": "Shambles",
    },
    Object {
      "address": "1467 Huzwas Highway",
      "id": "8242a55b-5a9d-5ca9-b296-eed08dcf2897",
      "rating": 0,
      "title": "Amigos",
    },
    Object {
      "address": "1915 Uzma Court",
      "id": "6bf9374f-8abc-5942-b52e-fff076137298",
      "rating": 0,
      "title": "Little China",
    },
    Object {
      "address": "1117 Nekli Glen",
      "id": "d1c878c1-1474-5c83-b289-508f75b89a4f",
      "rating": 0,
      "title": "The Hive",
    },
    Object {
      "address": "1006 Laet Manor",
      "id": "dc1884c7-35da-5973-9893-f902839ff6ff",
      "rating": 0,
      "title": "The Locket",
    },
    Object {
      "address": "1483 Sezsac Heights",
      "id": "ea7ace38-298a-556b-8395-48d49fd33c0d",
      "rating": 0,
      "title": "The Japanese Fence",
    },
    Object {
      "address": "1485 Lihom Boulevard",
      "id": "4416102f-3ff0-567f-adb8-bf3a5c353d9f",
      "rating": 0,
      "title": "The Bitter Window",
    },
    Object {
      "address": "1754 Lija Boulevard",
      "id": "d25bd8aa-ce6d-5936-8344-c767980bdac4",
      "rating": 0,
      "title": "The Fable Street",
    },
  ],
  "total": 20,
}
`;

exports[`List Places without any criteria should succeed 1`] = `
Object {
  "items": Array [
    Object {
      "address": "1187 Weskut Ridge",
      "id": "d3973942-133b-55e4-920e-7b0033ab9d38",
      "rating": 0,
      "title": "The Noodle Farmer",
    },
    Object {
      "address": "375 Uciwa Loop",
      "id": "55b62c9d-d3f6-5973-80cc-ca2744b0c1a6",
      "rating": 0,
      "title": "The Honor Goat",
    },
    Object {
      "address": "332 Ruvow Path",
      "id": "821ea0ca-77af-5dd5-a15f-176fd4ce23a2",
      "rating": 0,
      "title": "The Dairy Laguna",
    },
    Object {
      "address": "1253 Pusi Road",
      "id": "381424b3-ba66-5444-8d7a-a452071dc3b3",
      "rating": 0,
      "title": "The Bitter Window",
    },
    Object {
      "address": "1528 Aceer Way",
      "id": "36241277-6054-5955-a823-6a0706710235",
      "rating": 0,
      "title": "The Fire Grove",
    },
    Object {
      "address": "826 Cuzmav Ridge",
      "id": "f2ff6a02-8f86-51f7-bd0e-50099e450c27",
      "rating": 0,
      "title": "Shambles",
    },
    Object {
      "address": "1467 Huzwas Highway",
      "id": "8242a55b-5a9d-5ca9-b296-eed08dcf2897",
      "rating": 0,
      "title": "Amigos",
    },
    Object {
      "address": "1915 Uzma Court",
      "id": "6bf9374f-8abc-5942-b52e-fff076137298",
      "rating": 0,
      "title": "Little China",
    },
    Object {
      "address": "1117 Nekli Glen",
      "id": "d1c878c1-1474-5c83-b289-508f75b89a4f",
      "rating": 0,
      "title": "The Hive",
    },
    Object {
      "address": "1006 Laet Manor",
      "id": "dc1884c7-35da-5973-9893-f902839ff6ff",
      "rating": 0,
      "title": "The Locket",
    },
  ],
  "total": 20,
}
`;
